src/node_crypto.cc

```diff
@@ -4349,6 +4351,123 @@ void Hash::HashDigest(const FunctionCallbackInfo<Value>& args) {
+
+  CHECK(args[0]->IsInt32());
+  const int nid = args[0].As<Int32>()->Value();
+
+  CHECK(args[1]->IsArrayBufferView());
+  ByteSource key = ByteSource::FromBuffer(args[1]);
+
+  const EVP_MD* md = nullptr;
+  if (nid == EVP_PKEY_HMAC) {
+    CHECK(args[2]->IsString());
+    const node::Utf8Value name(env->isolate(), args[2]);
+    md = EVP_get_digestbyname(*name);
+    if (md == nullptr)
+      return CheckThrow(env, SignBase::Error::kSignUnknownDigest);
+  }
+
+  EVPKeyPointer key_ptr;
+  if (nid == EVP_PKEY_CMAC) {
+    CHECK(args[2]->IsString());
+    const node::Utf8Value name(env->isolate(), args[2]);
+    const EVP_CIPHER* cipher = EVP_get_cipherbyname(*name);
+    if (cipher == nullptr) return env->ThrowError("Unknown cipher");
+    key_ptr.reset(
+        EVP_PKEY_new_CMAC_key(nullptr,
+                              reinterpret_cast<const unsigned char*>(key.get()),
+                              key.size(), cipher));
+  } else {
+    key_ptr.reset(
+        EVP_PKEY_new_raw_private_key(
+            nid, nullptr, reinterpret_cast<const unsigned char*>(key.get()),
+            key.size()));
+  }
+
+  ManagedEVPPKey pkey(std::move(key_ptr));
+  EVPMDPointer mdctx(EVP_MD_CTX_new());
+
+
+  EVP_PKEY_CTX* pkctx = nullptr;
+  if (!EVP_DigestSignInit(mdctx.get(), &pkctx, md, nullptr, pkey.get()))
+    return ThrowCryptoError(env, ERR_get_error(), "EVP_DigestSignInit");
+
+  // TODO(bnoordhuis) Call EVP_PKEY_CTX_ctrl() or EVP_PKEY_CTX_ctrl_str().
+  USE(pkctx);
+
+  new Mac(env, args.This(), nid, std::move(mdctx));
+}
+
+
+void Mac::Update(const FunctionCallbackInfo<Value>& args) {
+  Decode<Mac>(args, [](Mac* mac, const FunctionCallbackInfo<Value>& args,
+                       const char* data, size_t size) {
+    CHECK_EQ(1, EVP_DigestSignUpdate(mac->mdctx_.get(), data, size));
+  });
+}
+
+
+void Mac::Final(const FunctionCallbackInfo<Value>& args) {
+  Mac* mac;
+  ASSIGN_OR_RETURN_UNWRAP(&mac, args.Holder());
+
+  MarkPopErrorOnReturn mark_pop_error_on_return;
+  Environment* env = Environment::GetCurrent(args);
+
+  enum encoding encoding = BUFFER;
+  if (args.Length() > 0)
+    encoding = ParseEncoding(env->isolate(), args[0], BUFFER);
+
+  size_t size;
+  if (!EVP_DigestSignFinal(mac->mdctx_.get(), nullptr, &size))
+    return ThrowCryptoError(env, ERR_get_error(), "EVP_DigestSignFinal");
+
+  unsigned char* data = MallocOpenSSL<unsigned char>(size);
+  ByteSource source =
+      ByteSource::Allocated(reinterpret_cast<char*>(data), size);
+
+  if (!EVP_DigestSignFinal(mac->mdctx_.get(), data, &size))
+    return ThrowCryptoError(env, ERR_get_error(), "EVP_DigestSignFinal");
+
+  Local<Value> error;
+  MaybeLocal<Value> rc =
+      StringBytes::Encode(env->isolate(), source.get(), source.size(),
+                          encoding, &error);
+
+  if (rc.IsEmpty()) {
+    CHECK(!error.IsEmpty());
+    env->isolate()->ThrowException(error);
+    return;
+  }
+
+  args.GetReturnValue().Set(rc.ToLocalChecked());
```

Comment:
```suggestion
  Local<Value> error;
  Local<Object> rc;
  
  if (!StringBytes::Encode(env->isolate(), source.get(), 
                           source.size(), encoding, &error)
                               .ToLocal(&rc)) {
      CHECK(!error.IsEmpty());
      env->isolate()->ThrowException(err);
      return;
  }
  args.GetReturnValue().Set(rc);  
```

Reply:
This is, for better or worse, the prevailing idiom in node_crypto.cc. I might do a whole-file update in a follow-up PR but not here.